### PR TITLE
Fix reactor initialization calling wrong JRuby Selector initialization

### PR DESCRIPTION
**Pull request by ning-github**
https://github.com/puma/puma/pull/3151 broke JRuby selector initialization via the following:

```
NIO::Selector.backends.delete(backend) => nil
```

What results is that `::NIO::Selector.new(nil)` is not the same as `::NIO::Selector.new`.

The former invokes
https://github.com/socketry/nio4r/blob/v2.5.2/ext/nio4r/org/nio4r/Selector.java#L47, which results in an exception being thrown.

This PR will restore the reactor initialization to call the intended https://github.com/socketry/nio4r/blob/v2.5.2/ext/nio4r/org/nio4r/Selector.java#L41

### Description
Please describe your pull request. Thank you for contributing! You're the best.

### Your checklist for this pull request

- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added (or updated) appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [ ] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [ ] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

### CODE REVIEW

**Comment by dentarg** [comment]
Can you add a test that catches the regression from #3151 ?

And solve this in a way that doesn't change the behavior for other platforms :) (see the failing test)

**Comment by ning-github** [comment]
Added a test to assert on proper initialization.

However, something peculiar about this behavior is that I'm only able to reproduce this issue in certain environments.

On my puma fork, I observe:
```
jruby-9.3.15.0 :001 > require 'nio'
 => true
jruby-9.3.15.0 :002 > NIO::Selector.new(nil)
 => #<NIO::Selector:0x43756cb>
```

However, for my actual application, I see the failure:
```
irb(main):001:0> require 'nio'
=> true
irb(main):002:0> NIO::Selector.new(nil)
Traceback (most recent call last):
       10: from /nix/store/zdb3yw8slcmxa8gdw8099c339kjdn2gb-jruby-9.3.15.0/bin/irb:25:in `<main>'
        9: from org/jruby/RubyKernel.java:1053:in `load'
        8: from /nix/store/zdb3yw8slcmxa8gdw8099c339kjdn2gb-jruby-9.3.15.0/lib/ruby/gems/shared/gems/irb-1.0.0/exe/irb:11:in `<main>'
        7: from org/jruby/RubyKernel.java:1238:in `catch'
        6: from org/jruby/RubyKernel.java:1238:in `catch'
        5: from org/jruby/RubyKernel.java:1508:in `loop'
        4: from org/jruby/RubyKernel.java:1092:in `eval'
        3: from (irb):2:in `evaluate'
        2: from org/jruby/RubyClass.java:890:in `new'
        1: from org/nio4r/Selector.java:49:in `initialize'
ArgumentError (:java is the only supported backend)

```

**Comment by ning-github** [comment]
FWIW, this is handled on the nio4r side as of v2.5.5 https://github.com/socketry/nio4r/blob/v2.5.5/ext/nio4r/org/nio4r/Selector.java#L46

**Comment by dentarg** [comment]
> However, something peculiar about this behavior is that I'm only able to reproduce this issue in certain environments.

Did you figure out the reason for this?

**Comment by ning-github** [comment]
> > However, something peculiar about this behavior is that I'm only able to reproduce this issue in certain environments.
> 
> Did you figure out the reason for this?

Yes, I figured out and it's actually ended up being quite simple :)
The only reason it was working on my local clone is because installation defaulted to the most recent version of `nio4r`, which is `2.7.4`, which means that it already included the NIO-side handling for this specific issue that was added in `2.5.5`.

When I pin the version [here](https://github.com/puma/puma/blob/v6.4.3/Gemfile#L8) in Puma's Gemfile to something below `2.5.5`, for example `2.5.2`, I see the `ArgumentError` that I'm making this change to address.

In summary, unless [this](https://github.com/puma/puma/blob/v6.4.3/Gemfile#L8) is bumped to declare that `2.5.5` is the oldest compatible version of `nio4r`, then this PR's change should be needed on the Puma side.

**Comment by dentarg** [comment]
Thanks for figuring that out! 

> In summary, unless [this](https://github.com/puma/puma/blob/v6.4.3/Gemfile#L8) is bumped to declare that `2.5.5` is the oldest compatible version of `nio4r`

It is in the in the gemspec the bump should be https://github.com/puma/puma/blob/271e4d0b697aabb022de1daa4a6f425f416cfafe/puma.gemspec#L17

Seeing 2.5.5 have been available for soon 4 years (see the below snippet from https://rubygems.org/gems/nio4r/versions)

```
2.5.5 February 05, 2021 (119 KB)
2.5.5 February 05, 2021 java (132 KB)
```

...this sounds like a pretty rare edge-case – I don't think we should do anything in Puma

I assume you can update nio4r in your application?

**Comment by dentarg** [comment]
I guess adjusting the gemspec doesn't hurt and make things technically more correct, so I wouldn't reject it.

**Comment by ning-github** [comment]
> I assume you can update nio4r in your application?

Yup, this is what we ended up doing.

> It is in the in the gemspec the bump should be

I finally got around to doing this, but it looks like `test_changing_nio4r_version_during_phased_restart` is now failing. Regardless, the point about this being a rare edge case is reasonable, and I'm fine either way with continuing to try to get this adjustment merged or just closing it at this point.

**Comment by nateberkopec** [comment]
Thanks for following up @ning-github - unless someone else wants to pipe in, I think we're good to say this is edge-casey enough to avoid making changes.
